File: asc/dashboards/page/asc_tracker/asc_tracker.py
import frappe

@frappe.whitelist()
def page_query():
    default_settings = frappe.get_doc('ASC Panel')
    default_year = default_settings.default_year
    # default_year = "2020-21"

    total_schools = frappe.db.sql("""select count(name) from `tabSchool` where enabled=1 """)[0][0]
    school_level = frappe.db.sql(""" select 
                                lt.name,
                                sum(case when lt.name = a.level then 1 else 0 end) as total                                 
                                from `tabLevel` lt,`tabASC` a                                    
                                where a.docstatus != 2 and a.year = '{0}'
                                group by lt.name
                                """.format(default_year))
    total_enrol = frappe.db.sql("""select SUM(IFNULL(urdu_medium_enrolment,0) + IFNULL(english_medium_enrolment,0) + IFNULL(sindhi_medium_enrolment,0)) from `tabASC` where year=%s and docstatus!=2""", (default_year))[0][0]
    total_staff = frappe.db.sql("""select SUM(IFNULL(total_teacher,0) + IFNULL(total_non_teaching_non_government_staff,0) + IFNULL(total_non_teaching_staff,0)) from `tabASC` where year=%s and docstatus!=2""", (default_year))[0][0]
    total_images= frappe.db.sql("""select count(*) from `tabASC Images` where year=%s""", (default_year))[0][0]
    
    stack =frappe.db.sql("""select 
                    sum(case when (year='{0}' and water_available='No' and docstatus!=2) then 1 else 0 end) as schools_no_water,
                    sum(case when (year='{0}' and electricity_connection='No Electricity Connection' and docstatus!=2) then 1 else 0 end) as schools_no_electricity,
                    sum(case when (year='{0}' and total_teacher=1 and docstatus!=2) then 1 else 0 end) as schools_single_teacher,
                    sum(case when (year='{0}' and availability_of_building='No' and docstatus!=2) then 1 else 0 end) as schools_no_building,
                    sum(case when (year='{0}' and docstatus!=2) then 1 else 0 end) as entered_asc_schools,
                    sum(case when (year='{0}' and status_detail='Functional' and docstatus!=2) then 1 else 0 end) as func_asc_schools,
                    sum(case when (year='{0}' and status_detail='Closed' and docstatus!=2) then 1 else 0 end) as closed_asc_schools
                    from `tabASC` 
                    where year='{0}' and docstatus!=2
                    """.format(default_year), as_dict=1)[0]
    district_asc = frappe.db.sql(""" select 
                        d.name,
                        sum(case when (a.district = d.name) then 1 else 0 end) as asc_count
                        from `tabASC` a, `tabDistrict` d
                        where a.docstatus!=2 and a.year=%s
                        group by d.name
                        """, (default_year))
    school_asc = frappe.db.sql(""" select 
                        sum(case when (s.district = d.name) then 1 else 0 end) as asc_count
                        from `tabSchool` s, `tabDistrict` d
                        group by d.name
                        """)
    district_images = frappe.db.sql(""" select 
                        sum(case when (a.district = d.name) then 1 else 0 end) as asc_count
                        from `tabASC Images` a, `tabDistrict` d
                        where a.docstatus!=2 and a.year=%s
                        group by d.name
                        """, (default_year))  
    dict_name=[]
    per_dict=[]
    img_per=[]
    chart_data=[]

    count = 0
    for x in district_asc:
        dict_name.append(x[0])
        for s in school_asc[count: count+1]:
            if s[0] != 0:
                v = x[1]/s[0]*100
                per_dict.append(round(v))
                break
        count+=1

    count = 0   
    for im in district_images:    
        for s in school_asc[count: count+1]:
            if s[0] != 0:
                i = im[0]/s[0]*100
                img_per.append(round(i))
                break
        count+=1

    count = 0
    for d in per_dict:    
        for img in img_per[count: count+1]:
            for name in dict_name[count: count+1]:
                row = [ d, img, name ]
                chart_data.append(row)
                break
            break
        count+=1
    chart_data.sort(key=lambda x: x[0], reverse=True)
    
    asc_per=[]
    image_per=[]
    districts=[]
    for xx in chart_data:
        asc_per.append(xx[0])
        image_per.append(xx[1])
        districts.append(xx[2])
    obj={
        "total_schools":total_schools,
        "total_staff":total_staff,
        "total_enrol":total_enrol,
        "total_images":total_images,
        "school_level":school_level,
        "stack":stack,
        "district_images":image_per,
        "district_names":districts,
        "per_dict":asc_per,
        "chart_data":chart_data,
    }
    return obj

File: asc/dashboards/page/asc_tracker/asc_tracker.js
frappe.pages['asc-tracker'].on_page_load = function (wrapper) {
  var page = frappe.ui.make_app_page({
    parent: wrapper,
    title: 'ASC Tracker',
    single_column: true
  });
  $(frappe.render_template("asc_tracker")).appendTo(page.main);
  frappe.call({
    method: "asc.dashboards.page.asc_tracker.asc_tracker.page_query",
    freeze: true,
    callback: function (r) {
      set.values(r)
      columnchart(r.message.district_names, r.message.per_dict, r.message.district_images);
      piechart(r.message.school_level);
    },
  });
  setInterval(async () => {
    await frappe.call({
      method: "asc.dashboards.page.asc_tracker.asc_tracker.page_query",
      callback: function (r) {
        set.values(r)
        columnchart(r.message.district_names, r.message.per_dict, r.message.district_images);
        piechart(r.message.school_level);
      },
    });
  }, 20000);
}
filters = {
  add: function (page) {
    // frappe.call({
    // 	method: "asc.dashboards.page.asc_tracker_dummy.asc_tracker_dummy.page_query",
    //   freeze:true,
    // 	callback: function (r) {
    //     console.log(r.message);
    // 		set.values(r)
    //     columnchart(r.message.district_names, r.message.per_dict, r.message.district_images);
    //     piechart(r.message.school_level);
    // 	},
    // });
  },
}
set = {
  values: function (r) {
    make_page.total_schools(r.message.total_schools);
    make_page.entered_asc_schools(r.message.stack.entered_asc_schools, r.message.total_schools);
    make_page.fun_asc_school(r.message.stack.func_asc_schools, r.message.stack.entered_asc_schools);
    make_page.close_asc_school(r.message.stack.closed_asc_schools, r.message.stack.entered_asc_schools);
    make_page.completed_asc(r.message.stack.entered_asc_schools);
    make_page.school_no_building(r.message.stack.schools_no_building, r.message.stack.entered_asc_schools);
    make_page.total_student_enrollment(r.message.total_enrol);
    make_page.school_no_electricity(r.message.stack.schools_no_electricity, r.message.stack.entered_asc_schools);
    make_page.total_single_teacher_school(r.message.stack.schools_single_teacher, r.message.stack.entered_asc_schools);
    make_page.school_no_water(r.message.stack.schools_no_water, r.message.stack.entered_asc_schools);
    make_page.total_staff(r.message.total_staff);
    make_page.total_upload_images(r.message.total_images);
  }
}
make_page = {
  total_schools: function (total_schools) {
    $("#total_schools").empty();
    $("#total_schools").html(total_schools.toLocaleString("en-US"));
  },
  entered_asc_schools: function (entered_asc_schools, total_schools) {
    $("#entered_asc_schools").empty();
    $("#entered_asc_schools").html(entered_asc_schools.toLocaleString("en-US"))
    const asc_per = entered_asc_schools / total_schools * 100
    const value = `(${asc_per.toFixed(0)}%)`
    $("#asc_percentage").empty();
    $("#asc_percentage").html(value);
  },
  fun_asc_school: function (fun_asc_school, total_schools) {
    $("#fun_asc_school").empty();
    $("#fun_asc_school").html(fun_asc_school.toLocaleString("en-US"));
    const fun_per = fun_asc_school / total_schools * 100
    const fun_value = `(${fun_per.toFixed(0)}%)`
    $("#fun_percentage").empty();
    $("#fun_percentage").html(fun_value);
  },
  close_asc_school: function (close_asc_school, total_schools) {
    $("#close_asc_school").empty();
    $("#close_asc_school").html(close_asc_school.toLocaleString("en-US"));
    const close_per = close_asc_school / total_schools * 100
    const close_value = `(${close_per.toFixed(0)}%)`
    $("#close_percentage").empty();
    $("#close_percentage").html(close_value);
  },
  completed_asc: function (completed_asc) {
    $("#completed_asc").empty();
    $("#completed_asc").html(completed_asc.toLocaleString("en-US"));
  },
  school_no_building: function (school_no_building, entered_asc_schools) {
    $("#school_no_building").empty();
    $("#school_no_building").html(school_no_building.toLocaleString("en-US"));
    const close_per = school_no_building / entered_asc_schools * 100
    const close_value = `(${close_per.toFixed(0)}%)`
    $("#no_building_percentage").empty();
    $("#no_building_percentage").html(close_value);
  },
  total_student_enrollment: function (total_student_enrollment) {
    $("#total_student_enrollment").empty();
    $("#total_student_enrollment").html(total_student_enrollment.toLocaleString("en-US"));
  },
  school_no_electricity: function (school_no_electricity, entered_asc_schools) {
    $("#school_no_electricity").empty();
    $("#school_no_electricity").html(school_no_electricity.toLocaleString("en-US"));
    const close_per = school_no_electricity / entered_asc_schools * 100
    const close_value = `(${close_per.toFixed(0)}%)`
    $("#no_electricity_percentage").empty();
    $("#no_electricity_percentage").html(close_value);
  },
  total_single_teacher_school: function (total_single_teacher_school, entered_asc_schools) {
    $("#total_single_teacher_school").empty();
    $("#total_single_teacher_school").html(total_single_teacher_school.toLocaleString("en-US"));
    const close_per = total_single_teacher_school / entered_asc_schools * 100
    const close_value = `(${close_per.toFixed(0)}%)`
    $("#single_teacher_percentage").empty();
    $("#single_teacher_percentage").html(close_value);
  },
  school_no_water: function (school_no_water, entered_asc_schools) {
    $("#school_no_water").empty();
    $("#school_no_water").html(school_no_water.toLocaleString("en-US"));
    const water_per = school_no_water / entered_asc_schools * 100
    const water_value = `(${water_per.toFixed(0)}%)`
    $("#no_water_percentage").empty();
    $("#no_water_percentage").html(water_value);
  },
  total_staff: function (total_staff) {
    $("#total_staff").empty();
    $("#total_staff").html(total_staff.toLocaleString("en-US"));
  },
  total_upload_images: function (total_images) {
    $("#total_upload_images").empty();
    $("#total_upload_images").html(total_images.toLocaleString("en-US"));
  },
}

function columnchart(district_names, district_asc, district_images) {
  Highcharts.chart('container1', {
    chart: {
      type: 'column'
    },
    title: {
      text: 'Data Entry Completion Status District Wise'
    },
    subtitle: {
      text: ''
    },
    xAxis: {
      categories: district_names,
      crosshair: true
    },
    yAxis: {
      min: 0,
      title: {
        text: ''
      }
    },
    tooltip: {
      headerFormat: '<span style="font-size:10px">{point.key}</span><table>',
      pointFormat: '<tr><td style="color:{series.color};padding:0">{series.name}: </td>' +
        '<td style="padding:0"><b>{point.y:.1f}%</b></td></tr>',
      footerFormat: '</table>',
      shared: true,
      useHTML: true
    },
    plotOptions: {
      column: {
        pointPadding: 0.2,
        borderWidth: 0,
        // stacking: 'percent',
        dataLabels: {
          enabled: true
        }
      }
    },
    series: [{
      name: 'ASC',
      data: district_asc
    }, {
      name: 'Images',
      data: district_images

    }]
  });
  // 	Highcharts.chart('container1', {
  //     chart: {
  //         type: 'column'
  //     },
  // 	  colors: ["#0162e8", "#00cccc", "#673ab7", "#00b9ff"],
  //     title: {
  //         text: 'Data entry Completion Status District wise',
  // 		align: 'left',
  // 		 x: 30,
  // 		  style: {

  //             fontWeight: 'bold'
  //         }
  //     },
  //      credits: {
  //       enabled: false,
  //     },
  //     exporting: {
  //       enabled: false,
  //     },
  // 	 plotOptions: {
  //         series: {
  //             pointWidth: 20
  //         }
  //     },
  //     subtitle: {
  //         text: ''
  //     },
  //     xAxis: {
  //         type: 'category',
  //         labels: {
  //             rotation: -45,
  //             style: {
  //                 fontSize: '12px',
  //                 fontFamily: 'Roboto',
  //             }
  //         }
  //     },
  //     yAxis: {
  //         min: 0,

  //     },
  //     legend: {
  //         enabled: false
  //     },
  //     tooltip: {
  //         pointFormat: 'Progress: <b>{point.y:.0f} Schools</b>'
  //     },
  //     series: [{
  //         name: 'ASC',
  //         data: district_asc,
  //         dataLabels: {
  //             enabled: true,
  //             rotation: -90,
  //             color: '#FFFFFF',
  //             align: 'right',
  //             format: '{point.y:.0f}', // one decimal
  //             y: 10, // 10 pixels down from the top
  //             style: {
  //                 fontSize: '10px',
  //                 fontFamily: 'Roboto',
  //             }
  //         }
  //     }]
  // });
}

function piechart(school_level) {
  // Make monochrome colors
  var pieColors = (function () {
    var colors = [],
      base = Highcharts.getOptions().colors[0],
      i;

    for (i = 0; i < 10; i += 1) {
      // Start out with a darkened base color (negative brighten), and end
      // up with a much brighter color
      colors.push(Highcharts.color(base).brighten((i - 2) / 7).get());
    }
    return colors;
  }());
  let data = []
  for (let i = 0; i < school_level.length; i++) {
    obj = {}
    const element = school_level[i];
    obj = { name: element[0], y: element[1] }
    data.push(obj)
  }
  // Build the chart
  Highcharts.chart('container2', {
    chart: {
      plotBackgroundColor: null,
      plotBorderWidth: null,
      plotShadow: false,
      type: 'pie'
    },

    credits: {
      enabled: false,
    },
    exporting: {
      enabled: false,
    },
    title: {
      text: ''

    },
    tooltip: {
      pointFormat: '{series.name}: <b>{point.percentage:.0f}%</b>'
    },
    accessibility: {
      point: {
        valueSuffix: '%'
      }
    },
    plotOptions: {
      pie: {
        allowPointSelect: true,
        cursor: 'pointer',
        colors: pieColors,
        center: [150, 98],
        dataLabels: {
          enabled: true,
          format: '<b>{point.name}</b><br>{point.percentage:.0f} %',
          distance: -50,
          filter: {
            property: 'percentage',
            operator: '>',
            value: 4
          }
        }
      }
    },
    series: [{
      name: 'Schools',
      data: data
    }]
  });
}
